File: src/aggregation.ts
import Supercluster from "supercluster";

import {
  createMapsAggregationRuntimeIndex,
  type MapsAggregationRuntimeFeature,
  type MapsAggregationRuntimeIndex,
} from "./aggregation-runtime";

export type MapMetricRecord = Record<string, number>;

export type MapPoint<TProperties = Record<string, unknown>> = {
  id?: string | number;
  label?: string;
  latitude: number;
  longitude: number;
  metrics?: MapMetricRecord;
  properties?: TProperties;
};

export type IndexedMapPoint<TProperties = Record<string, unknown>> = Required<MapPoint<TProperties>> & {
  id: string;
};

export type AggregatedMapPoint<TProperties = Record<string, unknown>> = {
  coordinates: [longitude: number, latitude: number];
  kind: "point";
  metrics: MapMetricRecord;
  point: IndexedMapPoint<TProperties>;
};

export type AggregatedMapCluster = {
  clusterId: number;
  coordinates: [longitude: number, latitude: number];
  expansionZoom: number;
  kind: "cluster";
  metrics: MapMetricRecord;
  pointCount: number;
  pointCountAbbreviated: string;
};

export type AggregatedMapFeature<TProperties = Record<string, unknown>> =
  | AggregatedMapCluster
  | AggregatedMapPoint<TProperties>;

export type ViewportAggregationQuery = {
  bounds: [west: number, south: number, east: number, north: number];
  zoom: number;
};

export type VisibleAggregationSummary = {
  bounds: ViewportAggregationQuery["bounds"];
  metrics: MapMetricRecord;
  visibleClusterCount: number;
  visiblePointCount: number;
  visibleUnclusteredCount: number;
  zoom: number;
};

export type ViewportAggregation<TProperties = Record<string, unknown>> = {
  features: Array<AggregatedMapFeature<TProperties>>;
  summary: VisibleAggregationSummary;
};

export type MapDensityViewportSummary = {
  bounds: ViewportAggregationQuery["bounds"];
  itemCount: number;
  kind: "map";
  metricKeys: string[];
  metrics: MapMetricRecord;
  visibleClusterCount: number;
  visiblePointCount: number;
  visibleUnclusteredCount: number;
  zoom: number;
};

export type MapPointFilter<TProperties = Record<string, unknown>> = (
  point: IndexedMapPoint<TProperties>,
) => boolean;

export type PointAggregationIndexOptions<TProperties = Record<string, unknown>> = {
  extent?: number;
  filterPoint?: MapPointFilter<TProperties>;
  maxZoom?: number;
  minZoom?: number;
  radius?: number;
};

export type PointAggregationIndex<TProperties = Record<string, unknown>> = {
  dispose(): void;
  getClusterExpansionZoom(clusterId: number): number;
  getClusterLeaves(
    clusterId: number,
    limit?: number,
    offset?: number,
  ): Array<IndexedMapPoint<TProperties>>;
  getPointById(pointId: string): IndexedMapPoint<TProperties> | null;
  getViewportAggregation(query: ViewportAggregationQuery): ViewportAggregation<TProperties>;
};

type SuperclusterPointProperties = {
  pointId: string;
  [metricKey: string]: number | string;
};

type SuperclusterClusterProperties = MapMetricRecord;

type SuperclusterFeature =
  | Supercluster.ClusterFeature<SuperclusterClusterProperties>
  | Supercluster.PointFeature<SuperclusterPointProperties>;

const compactNumberFormatter = new Intl.NumberFormat("en", {
  maximumFractionDigits: 1,
  notation: "compact",
});

export function createPointAggregationIndex<TProperties = Record<string, unknown>>(
  points: readonly MapPoint<TProperties>[],
  options: PointAggregationIndexOptions<TProperties> = {},
): PointAggregationIndex<TProperties> {
  const normalizedPoints = points
    .map((point, index) => normalizeMapPoint(point, index))
    .filter(isFiniteMapPoint)
    .filter((point) => options.filterPoint?.(point) ?? true);
  const pointLookup = new Map(normalizedPoints.map((point) => [point.id, point]));
  const runtimeIndex = createMapsAggregationRuntimeIndex(normalizedPoints, options);

  if (runtimeIndex) {
    return createRustPointAggregationIndex(runtimeIndex, pointLookup);
  }

  return createSuperclusterPointAggregationIndex(normalizedPoints, pointLookup, options);
}

export function createMapDensityViewportSummary<TProperties = Record<string, unknown>>(
  aggregation: ViewportAggregation<TProperties>,
): MapDensityViewportSummary {
  const metricRecords = aggregation.features.map((feature) => feature.metrics);
  const metricKeys = collectMapMetricKeys(metricRecords);

  return {
    bounds: aggregation.summary.bounds,
    itemCount: aggregation.summary.visiblePointCount,
    kind: "map",
    metricKeys,
    metrics: sumMapMetrics(metricRecords, metricKeys),
    visibleClusterCount: aggregation.summary.visibleClusterCount,
    visiblePointCount: aggregation.summary.visiblePointCount,
    visibleUnclusteredCount: aggregation.summary.visibleUnclusteredCount,
    zoom: aggregation.summary.zoom,
  };
}

export function getBoundsFromPoints<TProperties = Record<string, unknown>>(
  points: readonly MapPoint<TProperties>[],
): [west: number, south: number, east: number, north: number] | null {
  let west = Number.POSITIVE_INFINITY;
  let south = Number.POSITIVE_INFINITY;
  let east = Number.NEGATIVE_INFINITY;
  let north = Number.NEGATIVE_INFINITY;
  let hasPoint = false;

  for (const point of points) {
    if (!Number.isFinite(point.latitude) || !Number.isFinite(point.longitude)) {
      continue;
    }

    hasPoint = true;
    west = Math.min(west, point.longitude);
    south = Math.min(south, point.latitude);
    east = Math.max(east, point.longitude);
    north = Math.max(north, point.latitude);
  }

  return hasPoint ? [west, south, east, north] : null;
}

function createRustPointAggregationIndex<TProperties>(
  runtimeIndex: MapsAggregationRuntimeIndex,
  pointLookup: Map<string, IndexedMapPoint<TProperties>>,
): PointAggregationIndex<TProperties> {
  return {
    dispose() {
      runtimeIndex.dispose();
    },
    getClusterExpansionZoom(clusterId) {
      return runtimeIndex.getClusterExpansionZoom(clusterId);
    },
    getClusterLeaves(clusterId, limit = 10, offset = 0) {
      return runtimeIndex
        .getClusterLeaves(clusterId, limit, offset)
        .map((point) => requireIndexedPoint(pointLookup, point.id));
    },
    getPointById(pointId) {
      const point = runtimeIndex.getPointById(pointId);
      return point ? requireIndexedPoint(pointLookup, point.id) : null;
    },
    getViewportAggregation(query) {
      const aggregation = runtimeIndex.getViewportAggregation(query);

      return {
        features: aggregation.features.map((feature) =>
          toRustAggregatedMapFeature(feature, pointLookup),
        ),
        summary: aggregation.summary,
      };
    },
  };
}

function toRustAggregatedMapFeature<TProperties>(
  feature: MapsAggregationRuntimeFeature,
  pointLookup: Map<string, IndexedMapPoint<TProperties>>,
): AggregatedMapFeature<TProperties> {
  if (feature.kind === "cluster") {
    return {
      clusterId: feature.clusterId,
      coordinates: feature.coordinates,
      expansionZoom: feature.expansionZoom,
      kind: "cluster",
      metrics: feature.metrics,
      pointCount: feature.pointCount,
      pointCountAbbreviated: feature.pointCountAbbreviated,
    };
  }

  const point = requireIndexedPoint(pointLookup, feature.pointId);

  return {
    coordinates: [point.longitude, point.latitude],
    kind: "point",
    metrics: point.metrics,
    point,
  };
}

function requireIndexedPoint<TProperties>(
  pointLookup: Map<string, IndexedMapPoint<TProperties>>,
  pointId: string,
): IndexedMapPoint<TProperties> {
  const point = pointLookup.get(pointId);

  if (!point) {
    throw new Error(`Rust aggregation returned unknown point id ${JSON.stringify(pointId)}.`);
  }

  return point;
}

function createSuperclusterPointAggregationIndex<TProperties>(
  normalizedPoints: readonly IndexedMapPoint<TProperties>[],
  pointLookup: Map<string, IndexedMapPoint<TProperties>>,
  options: PointAggregationIndexOptions<TProperties>,
): PointAggregationIndex<TProperties> {
  const metricKeys = collectMapMetricKeys(normalizedPoints.map((point) => point.metrics));
  const tree = new Supercluster<SuperclusterPointProperties, SuperclusterClusterProperties>({
    extent: options.extent ?? 512,
    map: (properties) => mapSuperclusterProperties(properties, metricKeys),
    maxZoom: options.maxZoom ?? 16,
    minZoom: options.minZoom ?? 0,
    radius: options.radius ?? 72,
    reduce: (accumulated, properties) => {
      for (const metricKey of metricKeys) {
        accumulated[metricKey] =
          readNumericMetric(accumulated, metricKey) + readNumericMetric(properties, metricKey);
      }
    },
  });

  tree.load(
    normalizedPoints.map((point) => ({
      geometry: {
        coordinates: [point.longitude, point.latitude],
        type: "Point" as const,
      },
      properties: {
        pointId: point.id,
        ...point.metrics,
      },
      type: "Feature" as const,
    })),
  );

  return {
    dispose() {},
    getClusterExpansionZoom(clusterId) {
      return tree.getClusterExpansionZoom(clusterId);
    },
    getClusterLeaves(clusterId, limit = 10, offset = 0) {
      return tree
        .getLeaves(clusterId, limit, offset)
        .map((feature) => pointLookup.get(feature.properties.pointId))
        .filter(isDefined);
    },
    getPointById(pointId) {
      return pointLookup.get(pointId) ?? null;
    },
    getViewportAggregation(query) {
      const rawFeatures = getFeaturesForBounds(tree, query.bounds, query.zoom);
      const features = rawFeatures
        .map((feature) => toAggregatedMapFeature(feature, pointLookup, metricKeys, tree))
        .filter(isDefined);

      return {
        features,
        summary: summarizeMapFeatures(query, features, metricKeys),
      };
    },
  };
}

function normalizeMapMetrics(metrics: MapMetricRecord | undefined): MapMetricRecord {
  if (!metrics) {
    return {};
  }

  return Object.fromEntries(Object.entries(metrics).filter((entry) => Number.isFinite(entry[1])));
}

function collectMapMetricKeys(metricRecords: readonly (MapMetricRecord | undefined)[]): string[] {
  const metricKeys = new Set<string>();

  for (const metrics of metricRecords) {
    for (const metricKey of Object.keys(metrics ?? {})) {
      metricKeys.add(metricKey);
    }
  }

  return Array.from(metricKeys).sort((left, right) => left.localeCompare(right));
}

function sumMapMetrics(
  metricRecords: readonly (MapMetricRecord | undefined)[],
  metricKeys = collectMapMetricKeys(metricRecords),
): MapMetricRecord {
  const totals = Object.fromEntries(metricKeys.map((metricKey) => [metricKey, 0]));

  for (const metrics of metricRecords) {
    for (const metricKey of metricKeys) {
      totals[metricKey] += readNumericMetric(metrics ?? {}, metricKey);
    }
  }

  return totals;
}

function normalizeMapPoint<TProperties>(
  point: MapPoint<TProperties>,
  index: number,
): IndexedMapPoint<TProperties> {
  return {
    id: String(point.id ?? index),
    label: point.label ?? "",
    latitude: point.latitude,
    longitude: point.longitude,
    metrics: normalizeMapMetrics(point.metrics),
    properties: point.properties ?? ({} as TProperties),
  };
}

function isFiniteMapPoint<TProperties>(point: IndexedMapPoint<TProperties>) {
  return Number.isFinite(point.latitude) && Number.isFinite(point.longitude);
}

function mapSuperclusterProperties(
  properties: SuperclusterPointProperties,
  metricKeys: readonly string[],
): SuperclusterClusterProperties {
  const aggregated: SuperclusterClusterProperties = {};

  for (const metricKey of metricKeys) {
    aggregated[metricKey] = readNumericMetric(properties, metricKey);
  }

  return aggregated;
}

function getFeaturesForBounds(
  tree: Supercluster<SuperclusterPointProperties, SuperclusterClusterProperties>,
  bounds: ViewportAggregationQuery["bounds"],
  zoom: number,
): SuperclusterFeature[] {
  const roundedZoom = Math.round(zoom);

  if (bounds[0] <= bounds[2]) {
    return tree.getClusters(bounds, roundedZoom);
  }

  const features = [
    ...tree.getClusters([bounds[0], bounds[1], 180, bounds[3]], roundedZoom),
    ...tree.getClusters([-180, bounds[1], bounds[2], bounds[3]], roundedZoom),
  ];
  const seen = new Set<string>();

  return features.filter((feature) => {
    const key =
      feature.properties.cluster === true
        ? `cluster:${feature.properties.cluster_id}`
        : `point:${feature.properties.pointId}`;

    if (seen.has(key)) {
      return false;
    }

    seen.add(key);
    return true;
  });
}

function toAggregatedMapFeature<TProperties>(
  feature: SuperclusterFeature,
  pointLookup: Map<string, IndexedMapPoint<TProperties>>,
  metricKeys: readonly string[],
  tree: Supercluster<SuperclusterPointProperties, SuperclusterClusterProperties>,
): AggregatedMapFeature<TProperties> | null {
  const [longitude, latitude] = feature.geometry.coordinates;

  if (feature.properties.cluster === true) {
    const clusterId = feature.properties.cluster_id;
    const pointCount = feature.properties.point_count;
    const pointCountAbbreviated = feature.properties.point_count_abbreviated;

    if (
      typeof clusterId !== "number" ||
      !Number.isFinite(clusterId) ||
      typeof pointCount !== "number" ||
      !Number.isFinite(pointCount)
    ) {
      return null;
    }

    return {
      clusterId,
      coordinates: [longitude, latitude],
      expansionZoom: tree.getClusterExpansionZoom(clusterId),
      kind: "cluster",
      metrics: getMetricsFromProperties(feature.properties, metricKeys),
      pointCount,
      pointCountAbbreviated:
        typeof pointCountAbbreviated === "string"
          ? pointCountAbbreviated
          : compactNumberFormatter.format(pointCount),
    };
  }

  const pointId = feature.properties.pointId;
  if (typeof pointId !== "string") {
    return null;
  }

  const point = pointLookup.get(pointId);

  if (!point) {
    return null;
  }

  return {
    coordinates: [longitude, latitude],
    kind: "point",
    metrics: point.metrics,
    point,
  };
}

function getMetricsFromProperties(properties: Record<string, unknown>, metricKeys: readonly string[]) {
  const metrics: MapMetricRecord = {};

  for (const metricKey of metricKeys) {
    metrics[metricKey] = readNumericMetric(properties, metricKey);
  }

  return metrics;
}

function summarizeMapFeatures<TProperties>(
  query: ViewportAggregationQuery,
  features: readonly AggregatedMapFeature<TProperties>[],
  metricKeys: readonly string[],
): VisibleAggregationSummary {
  const metrics = Object.fromEntries(metricKeys.map((metricKey) => [metricKey, 0]));
  let visibleClusterCount = 0;
  let visiblePointCount = 0;
  let visibleUnclusteredCount = 0;

  for (const feature of features) {
    if (feature.kind === "cluster") {
      visibleClusterCount += 1;
      visiblePointCount += feature.pointCount;

      for (const metricKey of metricKeys) {
        metrics[metricKey] += feature.metrics[metricKey] ?? 0;
      }

      continue;
    }

    visiblePointCount += 1;
    visibleUnclusteredCount += 1;

    for (const metricKey of metricKeys) {
      metrics[metricKey] += feature.metrics[metricKey] ?? 0;
    }
  }

  return {
    bounds: query.bounds,
    metrics,
    visibleClusterCount,
    visiblePointCount,
    visibleUnclusteredCount,
    zoom: query.zoom,
  };
}

function readNumericMetric(properties: Record<string, unknown>, metricKey: string) {
  const value = properties[metricKey];

  return typeof value === "number" && Number.isFinite(value) ? value : 0;
}

function isDefined<TValue>(value: TValue | null | undefined): value is TValue {
  return value !== null && value !== undefined;
}

File: crates/maps-core/src/aggregation.rs
use std::{
    collections::{BTreeMap, BTreeSet},
    error::Error,
    fmt,
};

use geojson::{Feature, Geometry, JsonObject, Value as GeoJsonValue, feature::Id};
use serde_json::{Value as JsonValue, json};
use supercluster::{CoordinateSystem, Supercluster};

use crate::{IndexedMapPoint, MapCoordinate, MapMetricRecord};

const DEFAULT_EXTENT: f64 = 512.0;
const DEFAULT_MAX_ZOOM: u8 = 16;
const DEFAULT_MIN_ZOOM: u8 = 0;
const DEFAULT_RADIUS: f64 = 72.0;
const MIN_POINTS_PER_CLUSTER: u8 = 2;

/// Configuration for the Maps-owned point aggregation contract.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct PointAggregationOptions {
    pub extent: f64,
    pub max_zoom: u8,
    pub min_zoom: u8,
    pub radius: f64,
}

impl Default for PointAggregationOptions {
    fn default() -> Self {
        Self {
            extent: DEFAULT_EXTENT,
            max_zoom: DEFAULT_MAX_ZOOM,
            min_zoom: DEFAULT_MIN_ZOOM,
            radius: DEFAULT_RADIUS,
        }
    }
}

/// A viewport aggregation query using `[west, south, east, north]` bounds.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct ViewportAggregationQuery {
    pub bounds: [f64; 4],
    pub zoom: f64,
}

/// One unclustered map point visible in a viewport.
#[derive(Clone, Debug, PartialEq)]
pub struct AggregatedMapPoint {
    pub coordinates: [f64; 2],
    pub metrics: MapMetricRecord,
    pub point: IndexedMapPoint,
}

/// One map cluster visible in a viewport.
#[derive(Clone, Debug, PartialEq)]
pub struct AggregatedMapCluster {
    pub cluster_id: usize,
    pub coordinates: [f64; 2],
    pub expansion_zoom: usize,
    pub metrics: MapMetricRecord,
    pub point_count: usize,
    pub point_count_abbreviated: String,
}

/// A renderer-independent map feature returned by viewport aggregation.
#[derive(Clone, Debug, PartialEq)]
pub enum AggregatedMapFeature {
    Cluster(AggregatedMapCluster),
    Point(AggregatedMapPoint),
}

/// Summary of the points represented by the visible map features.
#[derive(Clone, Debug, PartialEq)]
pub struct VisibleAggregationSummary {
    pub bounds: [f64; 4],
    pub metrics: MapMetricRecord,
    pub visible_cluster_count: usize,
    pub visible_point_count: usize,
    pub visible_unclustered_count: usize,
    pub zoom: f64,
}

/// Renderer-independent viewport aggregation result.
#[derive(Clone, Debug, PartialEq)]
pub struct ViewportAggregation {
    pub features: Vec<AggregatedMapFeature>,
    pub summary: VisibleAggregationSummary,
}

/// Error produced when the Maps aggregation contract cannot be satisfied.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct PointAggregationError {
    message: String,
}

impl PointAggregationError {
    fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
        }
    }
}

impl fmt::Display for PointAggregationError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(&self.message)
    }
}

impl Error for PointAggregationError {}

/// Maps-owned point aggregation index.
///
/// The internal spatial hierarchy is an implementation detail. Maps owns all
/// public types, metric reduction, point lookup, antimeridian de-duplication,
/// and validation semantics around that hierarchy.
pub struct PointAggregationIndex {
    cluster_metric_cache: BTreeMap<usize, MapMetricRecord>,
    index: Supercluster,
    metric_keys: Vec<String>,
    point_lookup: BTreeMap<String, IndexedMapPoint>,
    source_point_count: usize,
}

impl PointAggregationIndex {
    /// Builds an immutable spatial hierarchy from normalized native map points.
    pub fn new(
        points: Vec<IndexedMapPoint>,
        options: PointAggregationOptions,
    ) -> Result<Self, PointAggregationError> {
        validate_options(options)?;

        let points = points
            .into_iter()
            .filter(|point| MapCoordinate::new(point.longitude, point.latitude).is_some())
            .map(sanitize_indexed_point)
            .collect::<Vec<_>>();
        let metric_keys = collect_metric_keys(&points);
        let point_lookup = points
            .iter()
            .cloned()
            .map(|point| (point.id.clone(), point))
            .collect::<BTreeMap<_, _>>();
        let features = points.iter().map(to_geojson_feature).collect::<Vec<_>>();
        let cluster_options = Supercluster::builder()
            .coordinate_system(CoordinateSystem::LatLng)
            .extent(options.extent)
            .max_zoom(options.max_zoom)
            .min_points(MIN_POINTS_PER_CLUSTER)
            .min_zoom(options.min_zoom)
            .radius(options.radius)
            .build();
        let mut index = Supercluster::new(cluster_options);
        index
            .load(features)
            .map_err(|error| PointAggregationError::new(error.to_string()))?;

        Ok(Self {
            cluster_metric_cache: BTreeMap::new(),
            index,
            metric_keys,
            point_lookup,
            source_point_count: points.len(),
        })
    }

    /// Returns a normalized point by stable id.
    #[must_use]
    pub fn get_point_by_id(&self, point_id: &str) -> Option<IndexedMapPoint> {
        self.point_lookup.get(point_id).cloned()
    }

    /// Returns leaf points for a cluster with pagination.
    pub fn get_cluster_leaves(
        &self,
        cluster_id: usize,
        limit: usize,
        offset: usize,
    ) -> Result<Vec<IndexedMapPoint>, PointAggregationError> {
        self.validate_cluster_id(cluster_id)?;

        Ok(self
            .index
            .get_leaves(cluster_id, limit, offset)
            .into_iter()
            .filter_map(|feature| point_id_from_feature(&feature))
            .filter_map(|point_id| self.point_lookup.get(&point_id).cloned())
            .collect())
    }

    /// Returns the zoom at which a cluster expands into multiple children.
    pub fn get_cluster_expansion_zoom(
        &self,
        cluster_id: usize,
    ) -> Result<usize, PointAggregationError> {
        self.validate_cluster_id(cluster_id)?;
        Ok(self.index.get_cluster_expansion_zoom(cluster_id))
    }

    /// Aggregates visible points and clusters for one viewport query.
    pub fn get_viewport_aggregation(
        &mut self,
        query: ViewportAggregationQuery,
    ) -> Result<ViewportAggregation, PointAggregationError> {
        validate_query(query)?;
        let zoom = rounded_zoom(query.zoom);
        let raw_features = self
            .index
            .get_clusters(query.bounds, zoom)
            .map_err(|error| PointAggregationError::new(error.to_string()))?;
        let mut seen = BTreeSet::new();
        let mut features = Vec::with_capacity(raw_features.len());

        for feature in raw_features {
            let Some(key) = feature_identity(&feature) else {
                continue;
            };
            if !seen.insert(key) {
                continue;
            }
            if let Some(feature) = self.resolve_aggregated_feature(&feature)? {
                features.push(feature);
            }
        }

        let summary = summarize_features(query, &features, &self.metric_keys);

        Ok(ViewportAggregation { features, summary })
    }

    fn resolve_aggregated_feature(
        &mut self,
        feature: &Feature,
    ) -> Result<Option<AggregatedMapFeature>, PointAggregationError> {
        let Some(coordinates) = coordinate_from_feature(feature) else {
            return Ok(None);
        };

        if is_cluster_feature(feature) {
            let Some(cluster_id) = cluster_id_from_feature(feature) else {
                return Ok(None);
            };
            let Some(point_count) = point_count_from_feature(feature) else {
                return Ok(None);
            };
            let expansion_zoom = self.get_cluster_expansion_zoom(cluster_id)?;
            let metrics = self.metrics_for_cluster(cluster_id)?;
            let point_count_abbreviated = feature
                .property("point_count_abbreviated")
                .map(json_value_to_string)
                .unwrap_or_else(|| point_count.to_string());

            return Ok(Some(AggregatedMapFeature::Cluster(AggregatedMapCluster {
                cluster_id,
                coordinates,
                expansion_zoom,
                metrics,
                point_count,
                point_count_abbreviated,
            })));
        }

        let Some(point_id) = point_id_from_feature(feature) else {
            return Ok(None);
        };
        let Some(point) = self.point_lookup.get(&point_id).cloned() else {
            return Ok(None);
        };

        Ok(Some(AggregatedMapFeature::Point(AggregatedMapPoint {
            coordinates: [point.longitude, point.latitude],
            metrics: point.metrics.clone(),
            point,
        })))
    }

    fn metrics_for_cluster(
        &mut self,
        cluster_id: usize,
    ) -> Result<MapMetricRecord, PointAggregationError> {
        if let Some(metrics) = self.cluster_metric_cache.get(&cluster_id) {
            return Ok(metrics.clone());
        }

        self.validate_cluster_id(cluster_id)?;
        let children = self
            .index
            .get_children(cluster_id)
            .map_err(|error| PointAggregationError::new(error.to_string()))?;
        let mut metrics = zero_metrics(&self.metric_keys);

        for child in children {
            if is_cluster_feature(&child) {
                let child_cluster_id = cluster_id_from_feature(&child).ok_or_else(|| {
                    PointAggregationError::new("cluster child is missing a numeric cluster id")
                })?;
                let child_metrics = self.metrics_for_cluster(child_cluster_id)?;
                add_metrics(&mut metrics, &child_metrics, &self.metric_keys);
            } else if let Some(point_id) = point_id_from_feature(&child)
                && let Some(point) = self.point_lookup.get(&point_id)
            {
                add_metrics(&mut metrics, &point.metrics, &self.metric_keys);
            }
        }

        self.cluster_metric_cache
            .insert(cluster_id, metrics.clone());
        Ok(metrics)
    }

    fn validate_cluster_id(&self, cluster_id: usize) -> Result<(), PointAggregationError> {
        if cluster_id < self.source_point_count {
            return Err(PointAggregationError::new(format!(
                "invalid cluster id {cluster_id}"
            )));
        }

        Ok(())
    }
}

fn validate_options(options: PointAggregationOptions) -> Result<(), PointAggregationError> {
    if !options.extent.is_finite() || options.extent <= 0.0 {
        return Err(PointAggregationError::new(
            "aggregation extent must be finite and greater than zero",
        ));
    }
    if !options.radius.is_finite() || options.radius < 0.0 {
        return Err(PointAggregationError::new(
            "aggregation radius must be finite and non-negative",
        ));
    }
    if options.min_zoom > options.max_zoom {
        return Err(PointAggregationError::new(
            "aggregation min zoom must not exceed max zoom",
        ));
    }
    if options.max_zoom == u8::MAX {
        return Err(PointAggregationError::new(
            "aggregation max zoom must be below 255",
        ));
    }

    Ok(())
}

fn validate_query(query: ViewportAggregationQuery) -> Result<(), PointAggregationError> {
    if !query.zoom.is_finite() || query.bounds.iter().any(|value| !value.is_finite()) {
        return Err(PointAggregationError::new(
            "viewport aggregation query must contain only finite values",
        ));
    }
    if query.bounds[1] > query.bounds[3] {
        return Err(PointAggregationError::new(
            "viewport south bound must not exceed north bound",
        ));
    }

    Ok(())
}

fn sanitize_indexed_point(mut point: IndexedMapPoint) -> IndexedMapPoint {
    point.metrics.retain(|_, value| value.is_finite());
    point
}

fn collect_metric_keys(points: &[IndexedMapPoint]) -> Vec<String> {
    points
        .iter()
        .flat_map(|point| point.metrics.keys().cloned())
        .collect::<BTreeSet<_>>()
        .into_iter()
        .collect()
}

fn to_geojson_feature(point: &IndexedMapPoint) -> Feature {
    let mut properties = JsonObject::new();
    properties.insert("pointId".to_owned(), json!(point.id));
    let [longitude, latitude] = javascript_supercluster_coordinate(point.longitude, point.latitude);

    Feature {
        bbox: None,
        geometry: Some(Geometry::new(GeoJsonValue::Point(vec![longitude, latitude]))),
        id: Some(Id::String(point.id.clone())),
        properties: Some(properties),
        foreign_members: None,
    }
}

/// Reproduces Supercluster v8's Float32 storage precision for projected input points.
///
/// The JavaScript implementation rounds initial Web Mercator x/y values through
/// `Math.fround` before clustering. The Rust dependency stores f64 values, so we
/// inverse-project the rounded values before loading them. This keeps the internal
/// hierarchy numerically compatible while source point coordinates remain exact in
/// `point_lookup` and in public unclustered-point results.
fn javascript_supercluster_coordinate(longitude: f64, latitude: f64) -> [f64; 2] {
    let projected_x = f64::from((longitude / 360.0 + 0.5) as f32);
    let sin = latitude.to_radians().sin();
    let projected_y =
        (0.5 - (0.25 * ((1.0 + sin) / (1.0 - sin)).ln()) / std::f64::consts::PI).clamp(0.0, 1.0);
    let projected_y = f64::from(projected_y as f32);
    let longitude = (projected_x - 0.5) * 360.0;
    let mercator_y = ((180.0 - projected_y * 360.0) * std::f64::consts::PI) / 180.0;
    let latitude = (360.0 * mercator_y.exp().atan()) / std::f64::consts::PI - 90.0;

    [longitude, latitude]
}

fn is_cluster_feature(feature: &Feature) -> bool {
    feature
        .property("cluster")
        .and_then(JsonValue::as_bool)
        .unwrap_or(false)
}

fn cluster_id_from_feature(feature: &Feature) -> Option<usize> {
    feature
        .property("cluster_id")
        .and_then(JsonValue::as_u64)
        .and_then(|value| usize::try_from(value).ok())
}

fn point_count_from_feature(feature: &Feature) -> Option<usize> {
    feature
        .property("point_count")
        .and_then(JsonValue::as_u64)
        .and_then(|value| usize::try_from(value).ok())
}

fn point_id_from_feature(feature: &Feature) -> Option<String> {
    match feature.id.as_ref()? {
        Id::String(value) => Some(value.clone()),
        Id::Number(value) => Some(value.to_string()),
    }
}

fn coordinate_from_feature(feature: &Feature) -> Option<[f64; 2]> {
    let geometry = feature.geometry.as_ref()?;
    let GeoJsonValue::Point(coordinates) = &geometry.value else {
        return None;
    };
    if coordinates.len() < 2 {
        return None;
    }

    MapCoordinate::new(coordinates[0], coordinates[1]).map(MapCoordinate::as_array)
}

fn feature_identity(feature: &Feature) -> Option<String> {
    if is_cluster_feature(feature) {
        return cluster_id_from_feature(feature).map(|cluster_id| format!("cluster:{cluster_id}"));
    }

    point_id_from_feature(feature).map(|point_id| format!("point:{point_id}"))
}

fn rounded_zoom(zoom: f64) -> u8 {
    zoom.round().clamp(0.0, f64::from(u8::MAX)) as u8
}

fn json_value_to_string(value: &JsonValue) -> String {
    match value {
        JsonValue::String(value) => value.clone(),
        JsonValue::Number(value) => value.to_string(),
        _ => value.to_string(),
    }
}

fn zero_metrics(metric_keys: &[String]) -> MapMetricRecord {
    metric_keys
        .iter()
        .cloned()
        .map(|metric_key| (metric_key, 0.0))
        .collect()
}

fn add_metrics(target: &mut MapMetricRecord, source: &MapMetricRecord, metric_keys: &[String]) {
    for metric_key in metric_keys {
        *target.entry(metric_key.clone()).or_insert(0.0) +=
            source.get(metric_key).copied().unwrap_or(0.0);
    }
}

fn summarize_features(
    query: ViewportAggregationQuery,
    features: &[AggregatedMapFeature],
    metric_keys: &[String],
) -> VisibleAggregationSummary {
    let mut metrics = zero_metrics(metric_keys);
    let mut visible_cluster_count = 0;
    let mut visible_point_count = 0;
    let mut visible_unclustered_count = 0;

    for feature in features {
        match feature {
            AggregatedMapFeature::Cluster(cluster) => {
                visible_cluster_count += 1;
                visible_point_count += cluster.point_count;
                add_metrics(&mut metrics, &cluster.metrics, metric_keys);
            }
            AggregatedMapFeature::Point(point) => {
                visible_point_count += 1;
                visible_unclustered_count += 1;
                add_metrics(&mut metrics, &point.metrics, metric_keys);
            }
        }
    }

    VisibleAggregationSummary {
        bounds: query.bounds,
        metrics,
        visible_cluster_count,
        visible_point_count,
        visible_unclustered_count,
        zoom: query.zoom,
    }
}
